bananas_server/application/bananas_server.py

```diff
@@ -0,0 +1,149 @@
+            dependencies=content_entry.dependencies,
+            tags=content_entry.tags,
+        )
+
+    def receive_PACKET_CONTENT_CLIENT_INFO_LIST(self, source, content_type, openttd_version):
+        version_major = (openttd_version >> 28) & 0xF
+        version_minor = (openttd_version >> 24) & 0xF
+        version_patch = (openttd_version >> 20) & 0xF
+        version = (version_major, version_minor, version_patch)
+
+        for content_entry in self._by_content_type[content_type]:
+            if content_entry.min_version and version < content_entry.min_version:
+                continue
+            if content_entry.max_version and version >= content_entry.max_version:
+                continue
+
+            self._send_content_entry(source, content_entry)
+
+    def receive_PACKET_CONTENT_CLIENT_INFO_EXTID(self, source, content_infos):
+        for content_info in content_infos:
+            content_entry = self._by_unique_id[content_info.content_type].get(content_info.unique_id)
+            if content_entry:
+                self._send_content_entry(source, content_entry)
+
+    def receive_PACKET_CONTENT_CLIENT_INFO_EXTID_MD5(self, source, content_infos):
+        for content_info in content_infos:
+            content_entry = (
+                self._by_unique_id_and_md5sum[content_info.content_type]
+                .get(content_info.unique_id, {})
+                .get(content_info.md5sum)
+            )
+            if content_entry:
+                self._send_content_entry(source, content_entry)
+
+    def receive_PACKET_CONTENT_CLIENT_INFO_ID(self, source, content_infos):
+        for content_info in content_infos:
+            content_entry = self._by_content_id.get(content_info.content_id)
+            if content_entry:
+                self._send_content_entry(source, content_entry)
+
+    def receive_PACKET_CONTENT_CLIENT_CONTENT(self, source, content_infos):
+        for content_info in content_infos:
+            content_entry = self._by_content_id[content_info.content_id]
+
+            try:
+                stream = self.storage.get_stream(content_entry)
+            except Exception:
+                log.exception("Error with storage, aborting for this client ...")
+                return
+
+            source.protocol.send_PACKET_CONTENT_SERVER_CONTENT(
+                content_type=content_entry.content_type,
+                content_id=content_entry.content_id,
+                filesize=content_entry.filesize,
+                filename=f"{content_entry.name} - {content_entry.version}",
+                stream=stream,
+            )
+
+    def get_by_content_id(self, content_id):
+        return self._by_content_id.get(content_id)
+
+    def get_by_unique_id_and_md5sum(self, content_type, unique_id, md5sum):
+        return self._by_unique_id_and_md5sum[content_type].get(unique_id, {}).get(md5sum)
+
+    def reload_md5sum_mapping(self):
+        for content_type in ContentType:
```

Comment:
Because the way Python works, this is not possible. These functions are not async (no `async` identifier in the function), and as such they will always execute from the start till the finish.

So there is no need to keep these things in mind. You got to love quirks in Python :D